lib/socketset/socketset.go

```diff
@@ -0,0 +1,218 @@
+// socket to recieve events from the kernel.
+package socketset
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"syscall"
+
+	"golang.org/x/sys/unix"
+
+	errwrap "github.com/pkg/errors"
+)
+
+// Shutdown closes the event file descriptor and unblocks receive by sending
+// a message to the pipe file descriptor. It must be called before close, and
+// should only be called once.
+func (obj *SocketSet) Shutdown() error {
+	// close the event socket so no more events are produced
+	if err := unix.Close(obj.fdEvents); err != nil {
+		return err
+	}
+	// send a message to the pipe to unblock select
+	return unix.Sendto(obj.fdPipe, nil, 0, &unix.SockaddrUnix{
+		Name: path.Join(obj.pipeFile),
+	})
+}
+
+// Close closes the pipe file descriptor. It must only be called after
+// shutdown has closed fdEvents, and unblocked receive. It should only be
+// called once.
+func (obj *SocketSet) Close() error {
+	if err := unix.Unlink(obj.pipeFile); err != nil {
+		return errwrap.Wrapf(err, "could not unbind %s", obj.pipeFile)
+	}
+	return unix.Close(obj.fdPipe)
+}
+
+// ReceiveBytes waits for bytes from fdEvents and parses them into a slice of
+// netlink messages. It will block until an event is produced, or shutdown
+// is called.
+func (obj *SocketSet) ReceiveBytes() ([]byte, error) {
+	// Select will return when any fd in fdSet (fdEvents and fdPipe) is ready
+	// to read.
+	_, err := unix.Select(obj.nfd(), obj.fdSet(), nil, nil, nil)
+	if err != nil {
+		// if a system interrupt is caught
+		if err == unix.EINTR { // signal interrupt
+			return nil, nil
+		}
+		return nil, errwrap.Wrapf(err, "error selecting on fd")
+	}
+	// receive the message from the netlink socket into b
+	b := make([]byte, os.Getpagesize())
+	n, _, err := unix.Recvfrom(obj.fdEvents, b, unix.MSG_DONTWAIT) // non-blocking receive
+	if err != nil {
+		// if fdEvents is closed
+		if err == unix.EBADF { // bad file descriptor
+			return nil, nil
+		}
+		return nil, errwrap.Wrapf(err, "error receiving messages")
+	}
+	// if we didn't get enough bytes for a header, something went wrong
+	if n < unix.NLMSG_HDRLEN {
+		return nil, fmt.Errorf("received short header")
+	}
+	b = b[:n] // truncate b to message length
+	return b, nil
+}
+
+// ReceiveNetlinkMessages is a wrapper around ReceiveBytes that returns a NetlinkMessage.
+func (obj *SocketSet) ReceiveNetlinkMessages() ([]syscall.NetlinkMessage, error) {
+	msgBytes, err := obj.ReceiveBytes()
+	if err != nil {
+		return nil, err
+	}
+	// use syscall to parse, as func does not exist in x/sys/unix
+	return syscall.ParseNetlinkMessage(msgBytes)
+}
+
+// UEvent struct which has attributes passed from a KOBJECT_NETWORK_UEVENT
+type UEvent struct {
+	// Default keys, as per https://github.com/torvalds/linux/blob/master/lib/kobject_uevent.c
+	Action    string
+	Devpath   string
+	Subsystem string
+
+	// Every other KV pair
+	Data map[string]string
+}
+
+// ReceiveUEvent is a wrapper around ReceiveBytes that returns a UEvent
```

Comment:
Doc strings and longer comments are full sentences and need to end with a period.